### PR TITLE
unbranded credential samples for Python

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

File: docs/samples/client/python/oauth/unbranded_auth_code_credential.md
Comment (timotheeguerin) [inline docs/samples/client/python/oauth/unbranded_auth_code_credential.md:1]: should those be python files?

File: docs/samples/client/python/oauth/unbranded_client_credential.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+from typing import Any, Optional
+from corehttp.credentials import AccessTokenInfo, TokenRequestOptions
+import time
+import httpx
+
+
+class UnbrandedClientCredential:
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/unbranded_client_credential.py:7]:
```suggestion
class OAuth2ClientCredential:
```

Don't use the word "unbranded", it's an internal term

File: docs/samples/client/python/oauth/unbranded_auth_code_credential.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        :paramtype options: TokenRequestOptions
+        :return: An access token for the specified scopes.
+        :rtype: ~corehttp.credentials.AccessTokenInfo
+        :raises ValueError: If no scopes are provided or if authority URL is missing
+        """
+        if not scopes:
+            raise ValueError("At least one scope must be provided.")
+
+        authority_url = self.authority_url
+        if options is not None:
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/unbranded_auth_code_credential.py:57]:
```suggestion
        # Options is None, if there was no auth information in the original TypeSpec
        if options is not None:
```

Don't hesitate to over-comment a sample

File: docs/samples/client/python/oauth/unbranded_auth_code_credential.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            auth_flows = getattr(options, "auth_flows", None)
+            if auth_flows and len(auth_flows) > 0:
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/unbranded_auth_code_credential.py]:
```suggestion
            # If there was at least one flow in the TypeSpec, pick the first one.
            # If your TypeSpec has several flows, you may want to loop to find the one you need
            if auth_flows:
```

File: docs/samples/client/python/oauth/oauth2_auth_code_credential.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        :rtype: ~corehttp.credentials.AccessTokenInfo
+        :raises ValueError: If no scopes are provided or if authority URL is missing
+        """
+        if not scopes:
+            raise ValueError("At least one scope must be provided.")
+
+        auth_flows = options.get("auth_flows")
+        # If there was at least one flow in the TypeSpec, pick the first one.
+        # If your TypeSpec has several flows, you may want to loop to find the one you need
+        if auth_flows and len(auth_flows) > 0:
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/oauth2_auth_code_credential.py:60]:
```suggestion
        if auth_flows:
```

You don't need to test it's not empty, that already done aat this point

File: docs/samples/client/python/oauth/oauth2_auth_code_credential.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            auth_flow = auth_flows[0]
+            authority_url = auth_flow.get("authorizationUrl")
+        authority_url = authority_url or self.authority_url
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/oauth2_auth_code_credential.py:63]: This will crash if there is flow, as the variable is never created
image

File: docs/samples/client/python/oauth/oauth2_client_credential.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                response = client.post(authority_url, auth=(self.client_id, self.client_secret), data=data)
+                
+                response.raise_for_status()
+                
+                token = response.json().get("access_token")
+                expires_in = response.json().get("expires_in")
+                expires_on = time.time() + expires_in
+                
+                return AccessTokenInfo(token, expires_on)
+        except httpx.HTTPError as ex:
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/oauth2_client_credential.py:70]: I would remove try/except and make the sample simpler

File: docs/samples/client/python/oauth/oauth2_client_credential.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            raise ex
+        except Exception as ex:
+            raise ValueError(f"Failed to get access token: {str(ex)}")
+
+
+credential = OAuth2ClientCredential(client_id="m2m", client_secret="secret")
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/oauth2_client_credential.py:76]: You should put an example of on to use it with a client. Just fake one like `MyClient` and show `MyClient(credential, endpoint)`

File: docs/samples/client/python/oauth/oauth2_client_credential.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+            token = response.json().get("access_token")
+            expires_in = response.json().get("expires_in")
+            expires_on = time.time() + expires_in
+
+            return AccessTokenInfo(token, expires_on)
+
+
+credential = OAuth2ClientCredential(client_id="m2m", client_secret="secret")
+access_token = credential.get_token_info(
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/oauth2_client_credential.py:77]: We should either remove that, or clarify that we don't expect customers to do that. It may be confusing, they may understand they need to call get_token themselves.

File: docs/samples/client/python/oauth/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      tokenUrl: "https://auth.example.com/token",
+      scopes: ["read", "write"],
+    }
+  ]>
+)
+```
+
+## Implement the OAuth2 Credential
+
+Follow the sample implementation to create a credential class that includes the `get_token_info` method, which returns an `AccessTokenInfo` object.
```
Comment (lmazuel) [inline docs/samples/client/python/oauth/README.md:27]: Can we link the ref doc of corehttp? https://learn.microsoft.com/en-us/python/api/overview/azure/corehttp-readme?view=azure-python-preview&preserve-view=true

**Review by lmazuel — APPROVED** [review]

**Review by timotheeguerin — APPROVED** [review]
